File: tests/fixtures/python/async-patterns.py
import asyncio


async def fetch_data(url: str) -> str:
    print("Fetching: " + url)
    await asyncio.sleep(1)
    return "response"


async def process_batch(items: list[str]) -> list[str]:
    results = []
    for item in items:
        result = await fetch_data(item)
        print("Got: " + result)
        results.append(result)
    return results


async def gather_results(urls: list[str]) -> list[str]:
    tasks = [fetch_data(url) for url in urls]
    results = await asyncio.gather(*tasks)
    print("All done")
    return list(results)


class AsyncService:
    def __init__(self, base_url: str):
        self.base_url = base_url

    async def get(self, path: str) -> str:
        url = self.base_url + path
        print("GET " + url)
        return await fetch_data(url)

    async def post(self, path: str, data: str) -> str:
        url = self.base_url + path
        print("POST " + url + " with " + data)
        return "created"

File: tests/fixtures/typescript/async-patterns.ts
// Async function with error handling
async function fetchJSON<T>(url: string): Promise<T> {
  console.log("Fetching: " + url);
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error("HTTP " + response.status);
  }
  return response.json();
}

// Promise chain
function processData(input: string): Promise<string> {
  return Promise.resolve(input)
    .then((data) => {
      console.log("Step 1: " + data);
      return data.toUpperCase();
    })
    .then((data) => {
      console.log("Step 2: " + data);
      return data.trim();
    });
}

// Async generator
async function* paginate<T>(
  fetcher: (page: number) => Promise<T[]>,
  maxPages: number = 10
): AsyncGenerator<T[], void, unknown> {
  for (let page = 0; page < maxPages; page++) {
    const items = await fetcher(page);
    if (items.length === 0) break;
    console.log("Page " + page);
    yield items;
  }
}

// Retry pattern
async function withRetry<T>(
  fn: () => Promise<T>,
  retries: number = 3,
  delay: number = 1000
): Promise<T> {
  for (let i = 0; i < retries; i++) {
    try {
      return await fn();
    } catch (err) {
      console.log("Retry " + (i + 1));
      if (i === retries - 1) throw err;
      await new Promise((r) => setTimeout(r, delay));
    }
  }
  throw new Error("unreachable");
}

// Concurrent execution
async function processAll(urls: string[]): Promise<string[]> {
  const results = await Promise.all(
    urls.map((url) => fetchJSON<string>(url))
  );
  console.log("All done: " + results.length);
  return results;
}

export { fetchJSON, processData, paginate, withRetry, processAll };
